// @drift-labs/sdk/lib/node/memcmp.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUserWithName = exports.getUserThatHasBeenLP = exports.getUserWithAuctionFilter = exports.getUserWithOrderFilter = exports.getNonIdleUserFilter = exports.getUserFilter = void 0;
const bs58_1 = __importDefault(require("bs58"));
const anchor_1 = require("@coral-xyz/anchor");
const userName_1 = require("./userName");
function getUserFilter() {
    return {
        memcmp: {
            offset: 0,
            bytes: bs58_1.default.encode(anchor_1.BorshAccountsCoder.accountDiscriminator('User')),
        },
    };
}
exports.getUserFilter = getUserFilter;
function getNonIdleUserFilter() {
    return {
        memcmp: {
            offset: 4350,
            bytes: bs58_1.default.encode(Uint8Array.from([0])),
        },
    };
}
exports.getNonIdleUserFilter = getNonIdleUserFilter;
function getUserWithOrderFilter() {
    return {
        memcmp: {
            offset: 4352,
            bytes: bs58_1.default.encode(Uint8Array.from([1])),
        },
    };
}
exports.getUserWithOrderFilter = getUserWithOrderFilter;
function getUserWithAuctionFilter() {
    return {
        memcmp: {
            offset: 4354,
            bytes: bs58_1.default.encode(Uint8Array.from([1])),
        },
    };
}
exports.getUserWithAuctionFilter = getUserWithAuctionFilter;
function getUserThatHasBeenLP() {
    return {
        memcmp: {
            offset: 4267,
            bytes: bs58_1.default.encode(Uint8Array.from([99])),
        },
    };
}
exports.getUserThatHasBeenLP = getUserThatHasBeenLP;
function getUserWithName(name) {
    return {
        memcmp: {
            offset: 72,
            bytes: bs58_1.default.encode(Uint8Array.from((0, userName_1.encodeName)(name))),
        },
    };
}
exports.getUserWithName = getUserWithName;
