// @drift-labs/sdk/lib/node/isomorphic/grpc.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createClient = exports.Client = exports.CommitmentLevel = exports.SubscribeUpdate = exports.SubscribeRequest = void 0;
const yellowstone_grpc_1 = __importDefault(require("@triton-one/yellowstone-grpc"));
exports.Client = yellowstone_grpc_1.default;
const yellowstone_grpc_2 = require("@triton-one/yellowstone-grpc");
Object.defineProperty(exports, "SubscribeRequest", { enumerable: true, get: function () { return yellowstone_grpc_2.SubscribeRequest; } });
Object.defineProperty(exports, "SubscribeUpdate", { enumerable: true, get: function () { return yellowstone_grpc_2.SubscribeUpdate; } });
Object.defineProperty(exports, "CommitmentLevel", { enumerable: true, get: function () { return yellowstone_grpc_2.CommitmentLevel; } });
// Export a function to create a new Client instance
function createClient(...args) {
    return new yellowstone_grpc_1.default(...args);
}
exports.createClient = createClient;
